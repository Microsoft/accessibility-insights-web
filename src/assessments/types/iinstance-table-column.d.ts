// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { IAssessmentInstanceRowData } from '../../DetailsView/components/assessment-instance-table';

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: IAssessmentInstanceRowData) => JSX.Element;
}
