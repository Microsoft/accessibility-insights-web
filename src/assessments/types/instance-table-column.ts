// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GeneratedAssessmentInstance } from 'common/types/store-data/assessment-result-data';

export interface AssessmentInstanceRowData<P = {}> {
    key: string;
    statusChoiceGroup: JSX.Element;
    visualizationButton?: JSX.Element;
    instance: GeneratedAssessmentInstance<P>;
}

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: AssessmentInstanceRowData) => JSX.Element;
}
