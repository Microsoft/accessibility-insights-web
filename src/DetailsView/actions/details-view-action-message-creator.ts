// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { autobind } from '@uifabric/utilities';

import { OnDetailsViewPivotSelected, SelectTestStepPayload } from '../../background/actions/action-payloads';
import { DevToolActionMessageCreator } from '../../common/message-creators/dev-tool-action-message-creator';
import { Messages } from '../../common/messages';
import { DetailsViewPivotType } from '../../common/types/details-view-pivot-type';
import { ManualTestStatus } from '../../common/types/manual-test-status';
import { DetailsViewRightContentPanelType } from '../components/left-nav/details-view-right-content-panel-type';
import {
    AddFailureInstancePayload,
    AssessmentActionInstancePayload,
    AssessmentToggleActionPayload,
    BaseActionPayload,
    ChangeAssessmentStepStatusPayload,
    ChangeInstanceSelectionPayload,
    ChangeInstanceStatusPayload,
    EditFailureInstancePayload,
    OnDetailsViewOpenPayload,
    RemoveFailureInstancePayload,
    SwitchToTargetTabPayload,
    ToggleActionPayload,
} from './../../background/actions/action-payloads';
import { FeatureFlagPayload } from './../../background/actions/feature-flag-actions';
import { TelemetryDataFactory } from './../../common/telemetry-data-factory';
import * as TelemetryEvents from './../../common/telemetry-events';
import { ExportResultType } from './../../common/telemetry-events';
import { VisualizationType } from './../../common/types/visualization-type';
import { WindowUtils } from './../../common/window-utils';

const messages = Messages.Visualizations;

export class DetailsViewActionMessageCreator extends DevToolActionMessageCreator {
    private windowUtils: WindowUtils;

    constructor(postMessage: (message: IMessage) => void, tabId: number, telemetryFactory: TelemetryDataFactory, windowUtils: WindowUtils) {
        super(postMessage, tabId, telemetryFactory);
        this.windowUtils = windowUtils;
    }
    public updateIssuesSelectedTargets(selectedTargets: string[]): void {
        const payload: string[] = selectedTargets;
        const message: IMessage = {
            type: messages.Issues.UpdateSelectedTargets,
            tabId: this._tabId,
            payload,
        };

        this.dispatchMessage(message);
    }

    @autobind
    public closePreviewFeaturesPanel(): void {
        const type = Messages.PreviewFeatures.ClosePanel;
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: BaseActionPayload = {
            telemetry,
        };

        this.dispatchMessage({
            type: type,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public closeScopingPanel(): void {
        const type = Messages.Scoping.ClosePanel;
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: BaseActionPayload = {
            telemetry,
        };

        this.dispatchMessage({
            type: type,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public closeSettingsPanel(): void {
        const type = Messages.SettingsPanel.ClosePanel;
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: BaseActionPayload = {
            telemetry,
        };

        this.dispatchMessage({
            type: type,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public setFeatureFlag(featureFlagId: string, enabled: boolean, event: React.MouseEvent<HTMLElement>): void {
        const type = Messages.FeatureFlags.SetFeatureFlag;
        const telemetry = this.telemetryFactory.forFeatureFlagToggle(
            event,
            enabled,
            TelemetryEvents.TelemetryEventSource.DetailsView,
            featureFlagId,
        );
        const payload: FeatureFlagPayload = {
            feature: featureFlagId,
            enabled: enabled,
            telemetry: telemetry,
        };

        this.dispatchMessage({
            type: type,
            tabId: this._tabId,
            payload: payload,
        });
    }

    public exportResultsClicked(exportResultsType: ExportResultType, exportedHtml: string, event: React.MouseEvent<HTMLElement>): void {
        const telemetryData = this.telemetryFactory.forExportedHtml(
            exportResultsType,
            exportedHtml,
            event,
            TelemetryEvents.TelemetryEventSource.DetailsView,
        );

        this.sendTelemetry(TelemetryEvents.EXPORT_RESULTS, telemetryData);
    }

    @autobind
    public copyIssueDetailsClicked(event: React.MouseEvent<any>): void {
        const telemetryData = this.telemetryFactory.withTriggeredByAndSource(event, TelemetryEvents.TelemetryEventSource.DetailsView);
        this.sendTelemetry(TelemetryEvents.COPY_ISSUE_DETAILS, telemetryData);
    }

    public updateFocusedInstanceTarget(instanceTarget: string[]): void {
        const payload: string[] = instanceTarget;
        const message: IMessage = {
            type: messages.Issues.UpdateFocusedInstance,
            tabId: this._tabId,
            payload,
        };

        this.dispatchMessage(message);
    }

    public selectDetailsView(event: React.MouseEvent<HTMLElement>, type: VisualizationType, pivot: DetailsViewPivotType): void {
        const payload: OnDetailsViewOpenPayload = {
            telemetry: this.telemetryFactory.forSelectDetailsView(event, type),
            detailsViewType: type,
            pivotType: pivot,
        };

        this.dispatchMessage({
            type: messages.DetailsView.Select,
            tabId: this._tabId,
            payload,
        });
    }

    public selectTestStep(event: React.MouseEvent<HTMLElement>, selectedStep: string, type: VisualizationType): void {
        const payload: SelectTestStepPayload = {
            telemetry: this.telemetryFactory.forSelectTestStep(event, type, selectedStep),
            selectedStep: selectedStep,
            selectedTest: type,
        };

        this.dispatchMessage({
            type: Messages.Assessment.SelectTestStep,
            tabId: this._tabId,
            payload: payload,
        });
    }

    public sendPivotItemClicked(pivotKey: string, ev?: React.MouseEvent<HTMLElement>): void {
        const telemetry = this.telemetryFactory.forDetailsViewNavPivotActivated(ev, pivotKey);

        const payload: OnDetailsViewPivotSelected = {
            telemetry: telemetry,
            pivotKey: DetailsViewPivotType[pivotKey],
        };

        this.dispatchMessage({
            type: Messages.Visualizations.DetailsView.PivotSelect,
            tabId: this._tabId,
            payload: payload,
        });
    }

    @autobind
    public switchToTargetTab(event: React.MouseEvent<HTMLElement>): void {
        const telemetry = this.telemetryFactory.fromDetailsView(event);
        const payload: SwitchToTargetTabPayload = {
            telemetry,
        };
        this.dispatchMessage({
            type: Messages.Tab.Switch,
            tabId: this._tabId,
            payload,
        });
    }

    public startOverAssessment(event: React.MouseEvent<any>, test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.forAssessmentActionFromDetailsView(test, event);
        const payload: ToggleActionPayload = {
            test,
            step,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.StartOver,
            tabId: this._tabId,
            payload,
        });
    }

    public enableVisualHelper(test: VisualizationType, step: string, shouldScan = true, sendTelemetry = true): void {
        const telemetry = sendTelemetry ? this.telemetryFactory.forAssessmentActionFromDetailsViewNoTriggeredBy(test) : null;
        const payload: AssessmentToggleActionPayload = {
            test,
            step,
            telemetry,
        };

        this.dispatchMessage({
            type: shouldScan ? Messages.Assessment.EnableVisualHelper : Messages.Assessment.EnableVisualHelperWithoutScan,
            tabId: this._tabId,
            payload,
        });
    }

    public disableVisualHelpersForTest(test: VisualizationType): void {
        const payload: ToggleActionPayload = {
            test,
        };

        this.dispatchMessage({
            type: Messages.Assessment.DisableVisualHelperForTest,
            tabId: this._tabId,
            payload,
        });
    }

    public disableVisualHelper(test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: ToggleActionPayload = {
            test,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.DisableVisualHelper,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public changeManualTestStatus(status: ManualTestStatus, test: VisualizationType, step: string, selector: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: ChangeInstanceStatusPayload = {
            test,
            step,
            status,
            selector,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.ChangeStatus,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public changeManualTestStepStatus(status: ManualTestStatus, test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: ChangeAssessmentStepStatusPayload = {
            test,
            step,
            status,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.ChangeStepStatus,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public undoManualTestStatusChange(test: VisualizationType, step: string, selector: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: AssessmentActionInstancePayload = {
            test,
            step,
            selector,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.Undo,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public undoManualTestStepStatusChange(test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: ChangeAssessmentStepStatusPayload = {
            test: test,
            step: step,
            telemetry: telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.UndoChangeStepStatus,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public changeAssessmentVisualizationState(
        isVisualizationEnabled: boolean,
        test: VisualizationType,
        step: string,
        selector: string,
    ): void {
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: ChangeInstanceSelectionPayload = {
            test,
            step,
            isVisualizationEnabled,
            selector,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.ChangeVisualizationState,
            tabId: this._tabId,
            payload,
        });
    }

    public addFailureInstance(description: string, test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: AddFailureInstancePayload = {
            test,
            step,
            description,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.AddFailureInstance,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public removeFailureInstance(test: VisualizationType, step: string, id: string): void {
        const telemetry = this.telemetryFactory.forTestStepFromDetailsView(test, step);
        const payload: RemoveFailureInstancePayload = {
            test,
            step,
            id,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.RemoveFailureInstance,
            tabId: this._tabId,
            payload,
        });
    }

    public detailsViewOpened(selectedPivot: DetailsViewPivotType): void {
        const telemetryData = this.telemetryFactory.forDetailsViewOpened(selectedPivot);
        this.sendTelemetry(TelemetryEvents.DETAILS_VIEW_OPEN, telemetryData);
    }

    @autobind
    public editFailureInstance(description: string, test: VisualizationType, step: string, id: string): void {
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: EditFailureInstancePayload = {
            test,
            step,
            id,
            description,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.EditFailureInstance,
            tabId: this._tabId,
            payload,
        });
    }

    public passUnmarkedInstances(test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: ToggleActionPayload = {
            test,
            step,
            telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.PassUnmarkedInstances,
            tabId: this._tabId,
            payload,
        });
    }

    public changeAssessmentVisualizationStateForAll(isVisualizationEnabled: boolean, test: VisualizationType, step: string): void {
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: ChangeInstanceSelectionPayload = {
            test: test,
            step: step,
            isVisualizationEnabled: isVisualizationEnabled,
            selector: null,
            telemetry: telemetry,
        };

        this.dispatchMessage({
            type: Messages.Assessment.ChangeVisualizationStateForAll,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public continuePreviousAssessment(event: React.MouseEvent<any>): void {
        const telemetry = this.telemetryFactory.fromDetailsView(event);
        const payload: BaseActionPayload = {
            telemetry: telemetry,
        };
        this.dispatchMessage({
            type: Messages.Assessment.ContinuePreviousAssessment,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public startOverAllAssessments(event: React.MouseEvent<any>): void {
        const telemetry = this.telemetryFactory.fromDetailsView(event);
        const payload: BaseActionPayload = {
            telemetry: telemetry,
        };
        this.dispatchMessage({
            type: Messages.Assessment.StartOverAllAssessments,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public cancelStartOver(event: React.MouseEvent<any>, test: VisualizationType, requirement: string): void {
        const telemetry = this.telemetryFactory.forCancelStartOver(event, test, requirement);
        const payload: BaseActionPayload = {
            telemetry: telemetry,
        };
        this.dispatchMessage({
            type: Messages.Assessment.CancelStartOver,
            tabId: this._tabId,
            payload,
        });
    }

    @autobind
    public cancelStartOverAllAssessments(event: React.MouseEvent<any>): void {
        const telemetry = this.telemetryFactory.fromDetailsView(event);
        const payload: BaseActionPayload = {
            telemetry: telemetry,
        };
        this.dispatchMessage({
            type: Messages.Assessment.CancelStartOverAllAssessments,
            tabId: this._tabId,
            payload,
        });
    }

    public changeRightContentPanel(viewType: DetailsViewRightContentPanelType): void {
        const payload: DetailsViewRightContentPanelType = viewType;
        const message = {
            tabId: this._tabId,
            type: Messages.Visualizations.DetailsView.SetDetailsViewRightContentPanel,
            payload: payload,
        };

        this.dispatchMessage(message);
    }
}
