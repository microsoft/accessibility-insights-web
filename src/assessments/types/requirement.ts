// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IColumn } from '@fluentui/react';
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { HyperlinkDefinition } from 'common/types/hyperlink-definition';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import { VisualizationInstanceProcessorCallback } from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn, InstanceTableHeaderType } from './instance-table-data';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results for which this returns false will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    instanceTableHeaderType?: InstanceTableHeaderType;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
