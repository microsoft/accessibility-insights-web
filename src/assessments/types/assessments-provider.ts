// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { DictionaryStringTo } from 'types/common-types';
import { Assessment } from './iassessment';
import { Requirement } from './requirement';

export interface AssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(visualizationType: VisualizationType): boolean;
    forType(visualizationType: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    forRequirementKey(key: string): Readonly<Assessment>;
    getStep(visualizationType: VisualizationType, key: string): Readonly<Requirement>;
    getStepMap(
        visualizationType: VisualizationType,
    ): Readonly<DictionaryStringTo<Readonly<Requirement>>>;
}
