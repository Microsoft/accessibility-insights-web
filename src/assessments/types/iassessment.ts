// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RequirementOrdering } from '../../common/assessment/requirement';
import { IAssesssmentVisualizationConfiguration } from '../../common/configs/visualization-configuration-factory';
import { AnyExtension } from '../../common/extensibility/extension-point';
import { IAssessmentData } from '../../common/types/store-data/iassessment-result-data';
import { VisualizationType } from '../../common/types/visualization-type';
import { ContentPageComponent } from '../../views/content/content-page';
import { TestStep } from './test-step';

interface BaseAssessment {
    key: string;
    type: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    steps: TestStep[];
    featureFlag?: { required?: string[] };
    executeAssessmentScanPolicy?: (scheduleScan: (step: string) => void, data: IAssessmentData) => void;
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<IAssesssmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => IAssesssmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
