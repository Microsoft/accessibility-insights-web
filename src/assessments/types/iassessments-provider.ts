// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from '../../common/types/visualization-type';
import { Assessment } from './iassessment';
import { TestStep } from './test-step';

// tslint:disable-next-line:interface-name
export interface IAssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(type: VisualizationType): boolean;
    forType(type: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    getStep(type: VisualizationType, key: string): Readonly<TestStep>;
    getStepMap(type: VisualizationType): Readonly<IDictionaryStringTo<Readonly<TestStep>>>;
}
