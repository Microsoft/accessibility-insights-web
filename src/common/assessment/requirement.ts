// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { keys, sortBy } from 'lodash';

import { IAssessmentsProvider } from '../../assessments/types/iassessments-provider';
import { TestStep } from '../../assessments/types/test-step';
import { ManualTestStatusData, TestStepData } from '../types/manual-test-status';
import { VisualizationType } from '../types/visualization-type';

export type RequirementDefinition = TestStep;
export type RequirementData = TestStepData;

export type RequirementResult = {
    definition: RequirementDefinition;
    data: RequirementData;
};
export type RequirementOrdering = RequirementOrderPart | RequirementOrderPart[];
export type RequirementOrderPart = (result: RequirementResult) => string | number;

export function getRequirementsResults(
    provider: IAssessmentsProvider,
    type: VisualizationType,
    stepStatus: ManualTestStatusData,
): RequirementResult[] {
    const test = provider.forType(type);

    function result(key): RequirementResult {
        const definition = provider.getStep(type, key);
        const data = stepStatus[key];
        return { definition, data };
    }

    const unsorted = keys(stepStatus).map(result);

    return sortBy(unsorted, test.requirementOrder);
}
