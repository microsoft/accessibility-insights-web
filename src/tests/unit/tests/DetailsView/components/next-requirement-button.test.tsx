// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { DefaultButton } from '@fluentui/react';
import { Requirement } from 'assessments/types/requirement';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import {
    NextRequirementButton,
    NextRequirementButtonDeps,
    NextRequirementButtonProps,
} from 'DetailsView/components/next-requirement-button';
import { shallow } from 'enzyme';
import * as React from 'react';
import { IMock, Mock } from 'typemoq';
describe('NextRequirementButton', () => {
    let messageCreatorMock: IMock<DetailsViewActionMessageCreator>;
    let eventStub: React.MouseEvent<HTMLElement>;
    let props: NextRequirementButtonProps;

    beforeEach(() => {
        messageCreatorMock = Mock.ofType(DetailsViewActionMessageCreator);
        eventStub = {} as React.MouseEvent<HTMLElement>;
        props = {
            deps: {
                detailsViewActionMessageCreator: messageCreatorMock.object,
            } as NextRequirementButtonDeps,
            nextRequirement: {
                key: 'some requirement key',
            } as Requirement,
            currentTest: -1,
        };
    });

    it('renders', () => {
        const rendered = shallow(<NextRequirementButton {...props} />);
        expect(rendered.getElement()).toMatchSnapshot();
    });

    it('validate next requirement button', () => {
        messageCreatorMock
            .setup(mock =>
                mock.selectNextRequirement(eventStub, props.nextRequirement.key, props.currentTest),
            )
            .verifiable();

        const rendered = shallow(<NextRequirementButton {...props} />);
        rendered.find(DefaultButton).prop('onClick')(eventStub);
        messageCreatorMock.verifyAll();
    });
});
