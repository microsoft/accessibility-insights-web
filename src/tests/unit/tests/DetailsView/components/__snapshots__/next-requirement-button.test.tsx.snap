// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextRequirementButton renders 1`] = `
<DocumentFragment>
  <mock-customizeddefaultbutton>
    <span />
  </mock-customizeddefaultbutton>
</DocumentFragment>
`;

exports[`NextRequirementButton renders: CustomizedDefaultButton props 1`] = `
{
  "children": <span>
    <Memo
      ariaLabel="next"
      iconName="ChevronRight"
    />
  </span>,
  "className": undefined,
  "onClick": [Function],
  "text": undefined,
}
`;
